src/http.rs

```diff
@@ -507,14 +549,24 @@ pub type RawHeaderLine = (String, Vec<u8>);
 /// >                ; obsolete line folding
 /// >                ; see Section 3.2.4
 /// > ```
-pub fn read_header<R: Reader>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
+pub fn read_header<R: Read>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
+    macro_rules! byte (
```

Comment:
Yes, but just like Pico, it's stateless. It will update the `Request` object with all the info it could find, but it won't keep track of it's state (as Pico's developer points out, keeping state makes it slower). So you can read some bytes, call `parse()`, and if it came back as `Status::Impartial` (or `-2` in Pico), you can still inspect some properties. Perhaps you check the route and can determine it's a 404 early before having to parse any more. You should, however, read more bytes into the same buffer, and call `parse()` again. The parser will start from the beginning.

It's faster this way because: keeping state (and checking it every single byte) is slower, and in **most** cases, reading from a socket should read the entire head, so you shouldn't often get `Partial` parses.